### PR TITLE
Fixed #1502: Proxying of unannounced trailers

### LINKED ISSUE #1502: Reverse proxying GRPC calls does not work
### 1. What version of Caddy are you running (`caddy -version`)?
Caddy 0.9.5

### 2. What are you trying to do?
I am trying to use caddy as a reverse proxy to a GRPC server.

### 3. What is your entire Caddyfile?
See 7. below.

### 4. How did you run Caddy (give the full command and describe the execution environment)?
See 7. below.

### 5. What did you expect to see?
Working GRPC calls.

### 6. What did you see instead (give full error messages and/or log)?
No errors or log output of any kind. The grpc-server never receives anything from caddy.

The client only sees that the connection was closed by the (caddy) server before any sensible data could be sent or received.

I have sniffed the traffic using wireshark. The entire communication of one call consists of the following steps:
1. TCP SYN+ACK Client <-> Server
2. HTTP/2 SETTINGS package Client -> Server
3. TCP FIN+ACK Client <-> Server (FIN initiated by the sever)

It seems like caddy doesn't know what to do with a SETTINGS package or doesn't support something requested in that package and just terminates the connection.

I know my way around HTTP/1 a little but I don't really know anything about HTTP/2 or SETTINGS packages. The only thing I found out is that a SETTINGS package does not contain any request (GET/POST etc.) and thus no URL. So caddy can't know to which vserver/middleware it should route the package, thus the problem is most likely not *in* the proxy middleware but somewhere in the initial connection handling. I haven't looked at the caddy code yet though. I will try to come around to doing that soon.

I have researched GRPC proxing and caddy before and I found these related issues. I have created a new one though because I would consider them related but different enough from this to justify a new issue.

- #631 Apparently grpc proxing does not work out-of-the-box with go1.6
- #54 This seems to ask for a far larger amount of features. I don't need any fancy REST <-> GRPC converter, I simply want to have a caddy server on ports 80/443 and at the same time want to be able to have a grpc server on those ports.

### 7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?
I have created a little test suite in go, which mostly automates testing if GRPC works. See [this repo](https://github.com/targodan/caddyGrpcTest) for the source code, including a Caddyfile with which I tested this.

It provides a go package where you can run `go test -v` on and it will tell you if everything is working or not.

### PR DESCRIPTION
**Pull request by lhecker**
As per #1512 this PR only contains the necessary changes to fix #1502 and nothing more.
Sorry for closing the other PR and sorry for the delays - I'm having a kinda hard time in life right now and things like that are the result I guess. 😕

P.S.: I have no idea why one of the commit tests failed... Maybe a race condition? Because I haven't seen such a failure personally yet. <sup>With Rust that wouldn't have happened.</sup>

### CODE REVIEW

**Comment by mholt** [comment]
Thanks @lhecker. It's not a problem - take care of things you need to. I hope things look up for you soon!

I'm not sure about the failing test either, but it's something worth looking into! Even if this change doesn't go into this next release, it'll get in one of the next ones.

Know that I and many others (perhaps unknowingly) appreciate your work. I'll review this soon, but might also wait until you're ready to circle back around to it.

**Comment by lhecker** [comment]
@mholt My life is slowly getting back to normal and starting next month everything should finally be a bit more relaxed again. 😂

I just rebased this PR on the current master. So how about we take another look at this PR now? The new & shiny support for some glorious gRPC proxying has been waiting far too long already. 🙂

P.S.: When the tests on this PR originally failed after submitting this PR it I bet a race condition was at fault. Since the underlying issue might not have been fixed yet I think we should carefully look out for that in caddy in the future. I mean: You do run the tests with `-race` enabled but who knows if that catches the rather spurious bugs (e.g. those who only happen if caddy is preempted or interrupted at a critical moment etc.).

**Comment by lhecker** [comment]
@mholt I fixed the 2 issues you mentioned. The PR indeed was quite a bit messy, but I cleaned it up now (no TODOs anymore or removed code).

The issue with the test was - or rather "is" - that it's only a fake HTTP request test. Since we never actually serialize and submit `r` (the `http.Request`), but instead submit the fake request/response directly to the proxy's `ServeHTTP` we will never know if proxying of unannounced trailers actually works. That will only ever be the case if we do a proper E2E test over localhost.
Code wise on my side everything's ready for a proper E2E though, after I added the new `trailerTestStringReader` struct just now. 🙂

File: caddyhttp/proxy/reverseproxy.go [hunk 1]
```diff
@@ -318,30 +318,61 @@ func (rp *ReverseProxy) ServeHTTP(rw http.ResponseWriter, outreq *http.Request,
 		}
 		pooledIoCopy(backendConn, conn)
 	} else {
+		// NOTE:
+		//   Closing the Body involves acquiring a mutex, which is a
```
Comment (mholt) [inline caddyhttp/proxy/reverseproxy.go]: Why would this require a mutex? I'm trying to see where it would be closed in a different goroutine.

Comment (lhecker) [inline caddyhttp/proxy/reverseproxy.go]: The Go stdlib uses a Mutex for locking the `res.Body` when calling `Close()` on it. See [here](https://github.com/golang/go/blob/66b5a2f3f0b1d94f69763114a85a977f5bc0277a/src/net/http/transfer.go#L905-L906). I thought that having a bit additional logic in our code is a okay-ish cost considering that the above `defer` will never ever once do anything meaningful until that very rare occurence that our code panics.

Comment (mholt) [inline caddyhttp/proxy/reverseproxy.go]: Ah, I see, we don't want to call Close() more than once. Makes sense.

**Comment by mholt** [comment]
Feel free to submit small PRs with the other changes you had in #1512 if you still want to do those. :+1:

**Comment by lhecker** [comment]
Yeah I'll do that. 🙂 And if I have the opportunity to contribute something again afterwards it'd probably be something to integrate caddy with kubernetes or so... I don't really want to use linkerd for that. 😂 And maybe something to make the config file more generic - I honestly fear that the current format might become a mayor painpoint as soon as we accumulate more and more plugins which aren't necessarely strictly centered around HTTP or domain names at all.

**Comment by mholt** [comment]
Let's discuss it more, either in an issue, or I've also sent you an email with some information. Thanks again! :+1:

**Pull request merged**